File: seminars/seminar05/Account.java
/**
 * Copyright (c) 2009 ISP RAS.
 * 109004, A. Solzhenitsina, 25, Moscow, Russia.
 * All rights reserved.
 *
 * $Id$
 * Created on 08.10.2009
 *
 */

package root.account;

/**
 * @author Victor Kuliamin
 *
 */
public class Account
{
  boolean blocked = false;
  int bound = 1000000;
  int balance = 0;
  int maxCredit = -1000;

  public boolean deposit(int sum)
  {
    if(blocked) 
      return false;
    else if(sum < 0 || sum > bound)
      return false;
    else
    {
      balance += sum;
      return true;
    }
  }
  
  public boolean withdraw(int sum)
  {
    if(blocked)
      return false;
    else if(sum < 0 || sum > bound)
      return false;
    else if(balance <= maxCredit + sum)
      return false;
    else
    {
      balance -= sum;
      return true;
    }
  }

  public int getBalance()   
  { 
    return balance; 
  }
  
  public int getMaxCredit() 
  { 
    return -maxCredit; 
  }
  
  public boolean isBlocked()
  {
    return blocked;
  }

  public void block()
  {
    blocked = true;
  }
  
  public boolean unblock()
  {
    if(balance < maxCredit)
      return false;
    else
      blocked = false;
    
    return true;
  }
  
  public boolean setMaxCredit(int mc)
  {
    if(mc < -bound || mc > bound) 
      return false;
    else
      maxCredit = -mc;
    
    return true;
  }
}

File: seminars/seminar06/python/account.py
class Account:
    def __init__(self):
        self.__blocked: bool = False
        self.__bound: int = 1000000
        self.__balance: int = 0
        self.__max_credit: int = -1000

    def deposit(self, _sum: int) -> bool:
        if self.__blocked :
            return False 
        if _sum < 0 or _sum > self.__bound:
            return False
        self.__balance += _sum
        return True

    def withdraw(self, _sum: int) -> bool:
        if self.__blocked :
            return False
        if _sum < 0 or _sum > self.__bound :
            return False
        if self.__balance <= self.__max_credit + _sum:
            return False
        self.__balance -= _sum
        return True

    def get_balance(self) -> int:
        return self.__balance

    def get_max_credit(self) -> int:
        return -self.__max_credit

    def is_blocked(self) -> bool:
        return self.__blocked

    def block(self) -> None:
        self.__blocked = True

    def unblock(self) -> bool:
        if self.__balance < self.__max_credit:
            return False
        self.__blocked = False
        return True

    def set_max_credit(self, mc: int) -> bool:
        if abs(mc) > self.__bound:
            return False
        self.__max_credit = -mc
        return True
